str/cell-type-spec/master_cell_spec_runner.py

```diff
@@ -0,0 +1,308 @@
+                    0
+                ]
+            )
+        )
+
+        # Assign scenario based on whether sign matches
+        if secondary_effect_size_sign == main_effect_size_sign:
+            scenario_same_tr[cell_type] = 4
+        else:
+            scenario_same_tr[cell_type] = 5
+
+    return scenario_same_tr
+
+
+def process_same_egene(cell_types_with_same_egene, row_coeff, row_variantid, row_cell_type, ld_path):
+    """
+    Processes the scenario where multiple cell types have the same gene as signficant (but not necessarily the same TR as being associated with the gene). [ie classifies as either scenario 3/4/5]
+    """
+    scenario_same_egene = {}
+
+    # Get the sign of the main cell type's effect size
+    main_effect_size_sign = int(np.sign(row_coeff))
+
+    # Load the LD matrix for the current gene
+    egene_ld = pd.read_csv(f'{ld_path}/{cell_types_with_same_egene["gene_name"].iloc[0]}.csv')
+
+    for cell_type in cell_types_with_same_egene['cell_type'].unique():
+        if cell_type == row_cell_type:
+            continue  # Skip the main cell type
+
+        # Get the subset for this secondary cell type
+        subset = cell_types_with_same_egene[cell_types_with_same_egene['cell_type'] == cell_type]
+
+        # Filter LD matrix to only relevant variantids
+        filtered_ld = egene_ld[egene_ld['Unnamed: 0'].isin(subset['variantid'])]
+
+        # Safety check: skip if no matching LD data
+        if filtered_ld.empty or row_variantid not in filtered_ld.columns:
+            continue
+
+        # Find variant in subset with max LD to row_variantid
+        max_ld_value = filtered_ld[row_variantid].max()
+        max_row = filtered_ld.loc[filtered_ld[row_variantid].idxmax()]
+        max_variantid = max_row['Unnamed: 0']
+
+        if max_ld_value <= 0.5:
+            scenario_same_egene[cell_type] = 3  # Not in high LD
+        else:
+            # Get the sign of the secondary variant's effect
+            secondary_coeff = subset[subset['variantid'] == max_variantid]['coeff_meta_fixed'].values
+            if secondary_coeff.size == 0:
+                continue  # Safety check
+            secondary_sign = int(np.sign(secondary_coeff[0]))
+
+            if secondary_sign != main_effect_size_sign:
+                scenario_same_egene[cell_type] = 5  # Opposite sign
+            else:
+                scenario_same_egene[cell_type] = 4  # Same sign
+
+    return scenario_same_egene
+
+
+def reprocess_scenario_2(
+    row_cell_type, scenario_2_celltypes, egene_chrom, row_pos, row_end, egene, row_motif, meta_scen2_path
+):
+    """
+    Reprocesses scenario 2 to check if can be reclassified as scenario 4/5 based on meta-analysis p value, based on the criteria from Cuomo et al. 2025.
+
+    """
+    reprocess_scen2_dict = {}
+
+    # Load the meta results for the given cell type
+    meta_results = pd.read_csv(f'{meta_scen2_path}/{row_cell_type}/meta_results.tsv', sep='\t')
```

Comment:
Assuming the tsvs are not too large to fit into memory at the same time, I'd suggest extracting this into a cached function to avoid reloading each file multiple times:

```
@functools.cache
def _load_meta_results(path: str, cell_type: str) -> pd.DataFrame:
    return pd.read_csv(f'{path}/{cell_type}/meta_results.tsv', sep='\t')

...
    meta_results = _load_meta_results(meta_scen2_path, row_cell_type)
...
```